### PR TITLE
EMB-88 Blackhole Monitoring

### PR DESCRIPTION
**Pull request by jacobleonwoffenden**
Fulfils https://dsdmoj.atlassian.net/browse/EMB-88
Supersedes https://github.com/ministryofjustice/hmpps-ems-platform/pull/35

### CODE REVIEW

File: modules/blackhole-monitor/variables.tf [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+
+locals {
+  cloudwatch_event_rule_name = "${var.resource_name_prefix}-blackhole-monitor-rule"
+  cloudwatch_event_target_id = "${var.resource_name_prefix}-blackhole-monitor-target"
+  lambda_iam_policy_name     = "${var.resource_name_prefix}-blackhole-monitor-policy"
+  lambda_iam_role_name       = "${var.resource_name_prefix}-blackhole-monitor-role"
+  lambda_function_name       = "${var.resource_name_prefix}-blackhole-monitor"
+  lambda_function_layer_name = "${var.resource_name_prefix}-blackhole-monitor-layer"
```
Comment (chelnak) [inline modules/blackhole-monitor/variables.tf:19]: I don't think you have any layer requirements for this lambda.

File: modules/blackhole-monitor/lambda.tf [hunk 2]
```diff
@@ -0,0 +1,27 @@
+data "archive_file" "source" {
```
Comment (chelnak) [inline modules/blackhole-monitor/lambda.tf:1]: Consider using [this module](https://github.com/ministryofjustice/hmpps-ems-platform-terraform-modules/tree/main/modules/standard-lambda). It should reduce the amount of duplication we have.

File: modules/blackhole-monitor/src/monitor.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+import os
+
+cloudwatch = boto3.client('cloudwatch')
+ec2 = boto3.client('ec2')
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+def lambda_handler(event, context):
+  # Get transit gateways
+  getTransitGateways = ec2.describe_transit_gateways()
```
Comment (chelnak) [inline modules/blackhole-monitor/src/monitor.py:13]: variables should be snake_case

File: modules/blackhole-monitor/src/monitor.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+  # Loop over transit gateways
+  for i in getTransitGateways['TransitGateways']:
+    print("Transit Gateway: " + i['TransitGatewayId'])
+    # Loop over transit gateway route tables filtering on transit gateway ID
+    getTransitGatewayRouteTables = ec2.describe_transit_gateway_route_tables(
```
Comment (chelnak) [inline modules/blackhole-monitor/src/monitor.py:19]: variables should be snake_case

File: modules/blackhole-monitor/src/monitor.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+ec2 = boto3.client('ec2')
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+def lambda_handler(event, context):
+  # Get transit gateways
+  getTransitGateways = ec2.describe_transit_gateways()
+
+  # Loop over transit gateways
+  for i in getTransitGateways['TransitGateways']:
```
Comment (chelnak) [inline modules/blackhole-monitor/src/monitor.py:16]: for loop variables should be descriptive..

e.g `for gateway in getTransitGateway`

File: modules/blackhole-monitor/src/monitor.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            'Name': 'transit-gateway-id',
+            'Values': [
+                i['TransitGatewayId'],
+            ]
+        }
+      ]
+    )
+    # Search transit gateways
+    for j in getTransitGatewayRouteTables['TransitGatewayRouteTables']:
+      print("  Transit Gateway Route Table: " + j['TransitGatewayRouteTableId'])
```
Comment (chelnak) [inline modules/blackhole-monitor/src/monitor.py:31]: Swap `print` for `logger`

**Review by chelnak — CHANGES_REQUESTED** [review]
LGTM.

Only thing that needs addressing is the lambda module. We can refactor the lambda in the next task.

File: modules/blackhole-monitor/src/monitor.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+              ]
+          }
+        ]
+      )
+
+      if len(searchBlackholedTransitGatewayRoutes['Routes']) == 0:
+        print("    No Transit Gateway Attachements in blackholed state")
+      else:
+        for l in searchBlackholedTransitGatewayRoutes['Routes']:
+          print("    Transit Gateway Attachment (Blackholed) " + k['TransitGatewayAttachments'][0]['TransitGatewayAttachmentId'] + " Route: " + k['DestinationCidrBlock'])
```
Comment (jacobleonwoffenden) [inline modules/blackhole-monitor/src/monitor.py:92]: Needs fixing. Loops with `l` but uses `k` inside loop

File: modules/blackhole-monitor/variables.tf (hunk 1 above)
Comment (jacobleonwoffenden) [reply modules/blackhole-monitor/variables.tf]: https://github.com/ministryofjustice/hmpps-ems-platform-terraform-modules/pull/7/commits/70f0cb813cdc20ba21e9659765f37553642839d7

File: modules/blackhole-monitor/src/monitor.py (hunk 3 above)
Comment (jacobleonwoffenden) [reply modules/blackhole-monitor/src/monitor.py]: https://github.com/ministryofjustice/hmpps-ems-platform-terraform-modules/pull/7/commits/6e4db5c00662c9f7a1a3ccd32bb33a0348b831c8

File: modules/blackhole-monitor/src/monitor.py (hunk 4 above)
Comment (jacobleonwoffenden) [reply modules/blackhole-monitor/src/monitor.py]: https://github.com/ministryofjustice/hmpps-ems-platform-terraform-modules/pull/7/commits/6e4db5c00662c9f7a1a3ccd32bb33a0348b831c8

File: modules/blackhole-monitor/src/monitor.py (hunk 5 above)
Comment (jacobleonwoffenden) [reply modules/blackhole-monitor/src/monitor.py]: https://github.com/ministryofjustice/hmpps-ems-platform-terraform-modules/pull/7/commits/2c597cb6676cc92b7c42e7604a002e740e661e2b

File: modules/blackhole-monitor/src/monitor.py (hunk 6 above)
Comment (jacobleonwoffenden) [reply modules/blackhole-monitor/src/monitor.py]: https://github.com/ministryofjustice/hmpps-ems-platform-terraform-modules/pull/7/commits/1fb3fd60ac169b61a8244a6fda0cedab16c9b90a

File: modules/blackhole-monitor/src/monitor.py (hunk 7 above)
Comment (jacobleonwoffenden) [reply modules/blackhole-monitor/src/monitor.py]: https://github.com/ministryofjustice/hmpps-ems-platform-terraform-modules/pull/7/commits/2c597cb6676cc92b7c42e7604a002e740e661e2b

File: modules/blackhole-monitor/lambda.tf (hunk 2 above)
Comment (jacobleonwoffenden) [reply modules/blackhole-monitor/lambda.tf:1]: What's the best approach for this? Do I use standard-lambda as a source in this repo/module or from https://github.com/ministryofjustice/hmpps-ems-platform like you have with https://github.com/ministryofjustice/hmpps-ems-platform/blob/main/accounts/hmpps-iam/inactive-account-monitor/lambda.tf#L31

Comment (chelnak) [reply modules/blackhole-monitor/lambda.tf:1]: Yeah, I think thats the right way to go. And we should be building out that module as we go along.

What do you think?

File: modules/blackhole-monitor/src/monitor.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+def lambda_handler(event, context):
+  # Get transit gateways
+  get_transit_gateways = ec2.describe_transit_gateways()
+
+  # Loop over transit gateways
+  for transit_gateway in get_transit_gateways['TransitGateways']:
+    logger("Transit Gateway: " + transit_gateway['TransitGatewayId'])
```
Comment (chelnak) [inline modules/blackhole-monitor/src/monitor.py:17]: Should be `logger.info('')`

See https://github.com/ministryofjustice/hmpps-ems-platform-terraform-modules/blob/bb139f18b8a9dd22069a6439fd030014e26ff359/modules/vpn-monitor/src/monitor.py#L43 for ref

**Review by chelnak — APPROVED** [review]

**Pull request merged**